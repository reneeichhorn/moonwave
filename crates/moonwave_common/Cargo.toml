[package]
name = "moonwave_common"
version = "0.1.0"
authors = ["Rene Eichhorn <rene.eichhorn1@gmail.com>"]
edition = "2018"

[dependencies]
nalgebra = "0.18"
bytemuck = "1.5"
lazy_static = "1.4"