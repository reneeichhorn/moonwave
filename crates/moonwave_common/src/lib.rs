pub use nalgebra::*;

mod color;
pub use color::*;
