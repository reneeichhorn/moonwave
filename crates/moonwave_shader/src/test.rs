use crate::*;
use moonwave_common::{Vector2, Vector3, Vector4};

mod moonwave_shader {
  pub use crate::*;
}

#[vertex]
struct SampleVertex {
  position: Vector3<f32>,
  uv: Vector2<f32>,
}

#[test]
fn test_basic_shader() {
  let mut shader = Shader::new()
    .add_color_output("color", ShaderType::Float4)
    .add_color_output("normal", ShaderType::Float4)
    .add_vertex_attributes::<SampleVertex>()
    .finish();

  // Build graph
  {
    let bootstrap_start = shader.get_bootstrap_node();
    let bootstrap_end = shader.get_bootstrap_end_node();

    let graph = shader.get_graph_mut();
    let const_node1 =
      graph.add_node_builder(base::create_constant_vec3(Vector3::new(0.0, 1.0, 2.0)));
    let const_node2 =
      graph.add_node_builder(base::create_constant_vec4(Vector4::new(1.0, 1.0, 1.0, 1.0)));
    let multiply = graph.add_node_builder(base::create_multiply(ShaderType::Float3));
    let extend = graph.add_node_builder(base::create_extend1(ShaderType::Float4, 1.0));

    graph
      .connect_name(bootstrap_start, "position", multiply, "x")
      .unwrap();
    graph
      .connect_name(const_node1, "output", multiply, "y")
      .unwrap();
    graph
      .connect_name(multiply, "output", extend, "input")
      .unwrap();
    graph
      .connect_name(extend, "output", bootstrap_end, "position")
      .unwrap();
    graph
      .connect_name(extend, "output", bootstrap_end, "color")
      .unwrap();
    graph
      .connect_name(const_node2, "output", bootstrap_end, "normal")
      .unwrap();
  }

  let out = shader.build_full();
  insta::assert_snapshot!("full_vs", out.0);
  insta::assert_snapshot!("full_fs", out.1);
  let color_out = shader.build(&[shader.get_color_output_entity("color")]);
  insta::assert_snapshot!("color_vs", color_out.0);
  insta::assert_snapshot!("color_fs", color_out.1);
  let normal_out = shader.build(&[shader.get_color_output_entity("normal")]);
  insta::assert_snapshot!("normal_vs", normal_out.0);
  insta::assert_snapshot!("normal_fs", normal_out.1);
}
