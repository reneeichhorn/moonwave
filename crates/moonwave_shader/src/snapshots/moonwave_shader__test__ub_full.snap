---
source: crates/moonwave_shader/src/test.rs
expression: built
---
BuiltShaderGraph {
    vs: "layout (location = 0) in vec3 a_position;\nlayout (location = 1) in vec2 a_uv;\nlayout (location = 0) out vec4 vs_var_10_0;\nlayout (binding = 0) uniform sample_uniform_a_block {\n\tfloat sample;\n\tvec3 position;\n} sample_uniform_a;\nlayout (binding = 1) uniform sample_uniform_b_block {\n\tfloat sample;\n} sample_uniform_b;\nvoid main() {\nfloat var_5_0 = float(1.0000000);\nfloat var_4_0 = sample_uniform_b.sample;\nfloat var_3_0 = sample_uniform_a.sample;\nvec3 var_3_1 = sample_uniform_a.position;\nvar_0_0 = a_position;\nvar_0_1 = a_uv;\nvec3 var_6_0 = var_0_0 * var_3_1;\nvec3 var_7_0 = var_6_0 * var_4_0;\nvec4 var_9_0 = var_7_0[0];\nvec4 var_9_1 = var_7_0[1];\nvec4 var_9_2 = var_7_0[2];\nvec4 var_9_3 = var_7_0[3];\nvec4 var_10_0 = vec4(var_9_0,var_9_1,var_9_2,var_5_0);\ngl_Position = var_10_0;\nvs_var_10_0 = var_10_0;\n}\n",
    fs: "layout (location = 0) in vec4 vs_var_10_0;\nlayout (location = 0) out vec4 f_color;\nlayout (binding = 1) uniform sample_uniform_b_block {\n\tfloat sample;\n} sample_uniform_b;\nvoid main() {\nvec4 var_10_0 = vs_var_10_0;\nfloat var_4_0 = sample_uniform_b.sample;\nvec3 var_8_0 = var_10_0 * var_3_0;\nf_color = var_8_0;\n}\n",
    required_uniforms: [
        BuiltUniform {
            binding: 0,
            name: "sample_uniform_a",
            attributes: [
                (
                    "sample",
                    Float,
                ),
                (
                    "position",
                    Float3,
                ),
            ],
            in_vs: true,
            in_fs: false,
        },
        BuiltUniform {
            binding: 1,
            name: "sample_uniform_b",
            attributes: [
                (
                    "sample",
                    Float,
                ),
            ],
            in_vs: true,
            in_fs: true,
        },
    ],
}
